manifests/database/postgresql.pp

```diff
@@ -10,20 +13,111 @@
+        }
+      }
+      postgresql::server::db { $dbname:
+        user     => $::foreman::db_username,
+        password => $password,
+        owner    => $::foreman::db_username,
+        encoding => 'utf8',
+        locale   => 'en_US.utf8',
+      }
+    } elsif $::foreman::db_node_type == 'slave' {
+      file { 'Write .pgpass':
+        content => "*:*:*:${::foreman::db_replication_username}:${::foreman::db_replication_password}\n",
+        path    => "${::foreman::params::postgres_home}/.pgpass",
+        owner   => $::postgresql::params::user,
+        group   => $::postgresql::params::group,
+        mode    => '0600',
+      }
+      ->
+      exec { 'Initialize new postgresql cluster':
+        command => "${::postgresql::params::bindir}/pg_basebackup -w -c fast -X stream -h ${::foreman::db_host} -p ${port} -U ${::foreman::db_replication_username} -D ${::postgresql::params::datadir}",
+        creates => "${::postgresql::params::datadir}/base",
+        user    => $::postgresql::params::user,
+      }
+      ->
+      class {'postgresql::server':
+        listen_addresses     => '*',
+        manage_recovery_conf => true,
+        needs_initdb         => false,
+      }
+      
+      postgresql::server::recovery { 'Configure recovery.conf':
+        standby_mode => 'on',
+        primary_conninfo => "host=${::foreman::db_host} port=${port} user=${::foreman::db_replication_username} password=${::foreman::db_replication_password}",
+        trigger_file => "${::postgresql::params::datadir}/failover",
+      }
+      postgresql::server::role { $::foreman::db_username:
+        password_hash => $password,
+      }
+
+    } else {
+      fail("db_node_type should be set to 'master' or 'slave' if db_cluster is true")
+    }
+    ::foreman::database::postgresql::db_access { $db_all_cluster_hostnames: }
+
+    postgresql::server::role { $::foreman::db_replication_username:
+      password_hash => $rep_password,
+      replication   => true,
+    }
+    postgresql::server::config_entry { 'wal_level':
+      value => 'hot_standby',
+    }
+    postgresql::server::config_entry { 'max_wal_senders':
+      value => count($db_all_cluster_hostnames),
+    }
+    postgresql::server::config_entry { 'wal_keep_segments':
+      value => '256',
+    }
+    postgresql::server::config_entry { 'hot_standby':
+      value => 'on',
+    }
 
-  postgresql::server::db { $dbname:
-    user     => $::foreman::db_username,
-    password => $password,
-    owner    => $::foreman::db_username,
-    encoding => 'utf8',
-    locale   => 'en_US.utf8',
+  } else {
+    class {'::postgresql::client':}
+    class {'::postgresql::server':}
+
+    postgresql::server::db { $dbname:
+      user     => $::foreman::db_username,
+      password => $password,
+      owner    => $::foreman::db_username,
+      encoding => 'utf8',
+      locale   => 'en_US.utf8',
+    }
   }
 
-  Postgresql::Server::Role[$::foreman::db_username] -> Postgresql::Server::Database[$dbname]
```

Comment:
This chaining was added in puppetlabs-postgresql 4.1.0. Please also increase the minimum version in metadata.json.